pkg/pgmodel/migrate.go

```diff
@@ -5,336 +5,186 @@
+// i.e. relying on the required Promscale database extension to do
+// the migrations
+func FirstExtVersionWithNewMigration() semver.Version {
+	return semver.MustParse(firstExtVersionWithNewMigration)
 }
 
-// CheckDependencies makes sure all project dependencies, including the DB schema
-// the extension, are set up correctly. This will set the ExtensionIsInstalled
-// flag and thus should only be called once, at initialization.
-func CheckDependencies(db *pgx.Conn, versionInfo VersionInfo, migrationFailedDueToLockError bool, extOptions extension.ExtensionMigrateOptions) (err error) {
-	if err = CheckSchemaVersion(context.Background(), db, versionInfo, migrationFailedDueToLockError); err != nil {
-		return err
-	}
-	return extension.CheckVersions(db, migrationFailedDueToLockError, extOptions)
-}
+func Migrate(conn *pgx.Conn, appVersion VersionInfo, leaseLock *util.PgAdvisoryLock, extOptions extension.ExtensionMigrateOptions) error {
+	// At startup migrators attempt to grab the schema-version lock. If this
+	// fails that means some other connector is running. All is not lost: some
+	// other connector may have migrated the DB to the correct version. We warn,
+	// then start the connector as normal. If we are on the wrong version, the
+	// normal version-check code will prevent us from running.
 
-// CheckSchemaVersion checks the DB schema version without checking the extension
-func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo, migrationFailedDueToLockError bool) error {
-	expectedVersion := semver.MustParse(versionInfo.Version)
-	dbVersion, err := getSchemaVersionOnConnection(ctx, conn)
-	if err != nil {
-		return fmt.Errorf("failed to check schema version: %w", err)
-	}
-	if versionCompare := dbVersion.Compare(expectedVersion); versionCompare != 0 {
-		if versionCompare < 0 && migrationFailedDueToLockError {
-			return fmt.Errorf("Failed to acquire the migration lock to upgrade the schema version and unable to run with the old version. Please ensure that no other Promscale connectors with the old schema version are running. Received schema version %v but expected %v", dbVersion, expectedVersion)
+	if leaseLock != nil {
+		locked, err := leaseLock.GetAdvisoryLock()
+		if err != nil {
+			return fmt.Errorf("error while acquiring migration lock %w", err)
+		}
+		if !locked {
+			return MigrationLockError
 		}
-		return fmt.Errorf("Error while comparing schema version: received schema version %v but expected %v", dbVersion, expectedVersion)
+		defer func() {
+			_, err := leaseLock.Unlock()
+			if err != nil {
+				log.Error("msg", "error while releasing migration lock", "err", err)
+			}
+		}()
+	} else {
+		log.Warn("msg", "skipping migration lock")
 	}
-	return nil
-}
 
-type Migrator struct {
-	db       *pgx.Conn
-	sqlFiles http.FileSystem
-	toc      map[string][]string
-}
+	firstAppVersionWithNewMigration := FirstAppVersionWithNewMigration()
+	appSemver := semver.MustParse(appVersion.Version)
```

Comment:
I think we should be a bit conservative here and double check the extension we need is first available before doing any of this migration logic. Something like
```
_, _, err := extensionVersions(conn, "promscale", version.ExtVersionRange, version.ExtVersionRangeString, extOptions)
// check for err. If there is an err that means that the right extension version is unavailable
```